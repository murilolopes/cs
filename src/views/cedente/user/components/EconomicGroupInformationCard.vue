<template>
  <b-card :border-variant="borderVariant" :class="{ 'pending-documentation-bg-color': documentationInProgress }">
    <div class="d-flex flex-row justify-content-between align-items-start w-100">
      <div class="d-flex flex-column align-items-start justify-content-start w-50">
        <h3 class="mb-0" v-if="cedente.nome_fantasia.length > 20" tabindex="0" v-b-tooltip.hover :title="cedente.nome_fantasia">{{ truncateCedenteName }}</h3>
        <h3 class="mb-0" v-else>{{ cedente.nome_fantasia }}</h3>
        <p class="lead text-dark">{{ cedente.cnpj.replace(/^(\d{2})(\d{3})(\d{3})(\d{4})(\d{2})/, '$1.$2.$3/$4-$5') }}</p>
        <p class="lead text-muted">{{ subsidiarySizeLabel }}</p>
      </div>

      <div class="d-flex flex-row align-items-start justify-content-end w-50">
        <div class="d-flex flex-column justify-content-between align-items-end gap-1">
          <div class="d-flex flex-row align-items-center">
            <feather-icon :icon="badgeStatus.icon" size="26" :class="`text-${badgeStatus.variant}`" />
            <b-badge :variant="`light-${badgeStatus.variant}`" :class="`border-${badgeStatus.variant} ml-1`">{{
              badgeStatus.text }}</b-badge>
          </div>
          <div class="d-flex flex-row align-items-center">
            <b-button variant="primary" v-if="showAnalisyButton" @click="goToCedentePage">{{ buttonLabel }}</b-button>
          </div>
        </div>
      </div>
    </div>
  </b-card>
</template>

<script>
import {
  BCard,
  BBadge,
  BButton,
  VBTooltip
} from "bootstrap-vue";

export default {
  name: "EconomicGroupInformationCard",
  components: {
    BCard,
    BBadge,
    BButton,
  },
  props: {
    cedente: {
      type: Object,
      required: true,
      default: () => ({
        status: "nao_enviado",
        filiais: [],
      }),
    },
  },
  directives: {
    'b-tooltip': VBTooltip,
  },
  methods: {
    goToCedentePage() {
      const routeName = this.userType === 'cedente' ? 'complete-economic-group-registration' : 'admin-economicGroup.general-data';
      const params = this.userType === 'cedente' ? { id: this.cedente.id } : { id: this.cedente.grupo_economico_id, idMatriz: this.cedente.id };

      console.log(params)

      this.$router.push({ name: routeName, params });
    },
  },
  computed: {
    documentationInProgress() {
      return !['aprovado', 'reprovado'].includes(this.cedente.status_documentacao);
    },
    userType() {
      return this.$store.state.auth.userData.user_type
    },
    buttonLabel() {
      const label = {
        cedente: 'Preencher informações agora',
        admin: 'Analisar informações',
      }
      return label[this.userType]
    },
    showAnalisyButton() {
      const status = {
        cedente: {
          nao_enviado: true,
          em_validacao: false,
          aprovado: false,
          reprovado: true,
          blocked: false,
        },
        admin: {
          nao_enviado: true,
          em_validacao: true,
          aprovado: false,
          reprovado: true,
          blocked: true,
        },
      }

      return status[this.userType][this.cedente.status_documentacao]
    },
    subsidiarySizeLabel() {
      if (this.cedente.filiais?.length === 0) return "Sem filiais";

      return `+${this.cedente.filiais?.length} ${this.cedente.filiais?.length > 1 ? "Filiais incluídas" : "Filial incluída"}`;
    },
    borderVariant() {
      const status = {
        em_validacao: "light",
        nao_enviado: "light",
        reprovado: "danger",
        aprovado: "success",
        blocked: "light",
      }

      return status[this.cedente.status_documentacao]
    },
    badgeStatus() {
      const status = {
        cedente: {
          nao_enviado: {
            icon: 'AlertCircleIcon',
            variant: 'warning',
            text: 'Aguardando seu preenchimento',
          },
          em_validacao: {
            icon: 'ClockIcon',
            variant: 'warning',
            text: 'Informações em análise',
          },
          aprovado: {
            icon: 'CheckCircleIcon',
            variant: 'success',
            text: 'Aprovado',
          },
          reprovado: {
            icon: 'AlertOctagonIcon',
            variant: 'danger',
            text: 'Pendência no cadastro',
          },
          blocked: {
            icon: 'XCircleIcon',
            variant: 'dark',
            text: 'Conta temporariamente bloqueada',
          },
        },
        admin: {
          nao_enviado: {
            icon: 'AlertCircleIcon',
            variant: 'warning',
            text: 'Informações incompletas',
          },
          em_validacao: {
            icon: 'ClockIcon',
            variant: 'warning',
            text: 'Aguardando análise',
          },
          aprovado: {
            icon: 'CheckCircleIcon',
            variant: 'success',
            text: 'Aprovado',
          },
          reprovado: {
            icon: 'AlertOctagonIcon',
            variant: 'danger',
            text: 'Reprovado',
          },
          blocked: {
            icon: 'XCircleIcon',
            variant: 'dark',
            text: 'Oferta desativada',
          },
        }
      }

      return status[this.userType][this.cedente.status_documentacao]
    },
    truncateCedenteName() {
      return this.cedente.nome_fantasia.slice(0, this.cedente.nome_fantasia.length - 10).trim() + '...'
    },
  },
};
</script>

<style lang="scss">
.pending-documentation-bg-color {
  background-color: #F8FAFC !important;
}

.gap-1 {
  gap: 1.5rem;
}
</style>
