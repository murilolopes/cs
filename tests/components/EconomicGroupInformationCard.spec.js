import Vue from 'vue';
import BootstrapVue, { BButton, BCard, BBadge } from 'bootstrap-vue';
import FeatherIcon from '@core/components/feather-icon/FeatherIcon.vue';
import { shallowMount, mount } from '@vue/test-utils';
import EconomicGroupInformationCard from '@/views/cedente/user/components/EconomicGroupInformationCard.vue';

Vue.use(BootstrapVue);
Vue.component(FeatherIcon.name, FeatherIcon);

describe('EconomicGroupInformationCard.vue', () => {
  it('Should render correctly', () => {
    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup: {
          id: 1,
          name: 'Teste 1',
          cnpj: '12.345.678/0001-34',
          status: 'pending_documentation',
          filiais: [{ id: 1 }],
        },
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Should render correct subsidiary amount when does not have subsidiaries', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'pending_documentation',
      filiais: [],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    expect(wrapper.text()).toContain('Sem filiais');
  });

  it('Should render correct subsidiary amount when have just one subsidiary', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'pending_documentation',
      filiais: [{ id: 1 }, { id: 1 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    expect(wrapper.text()).toContain('+2 Filiais incluídas');
  });

  it('Should render correct subsidiary amount when having more than one subsidiary', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'pending_documentation',
      filiais: [{ id: 1 }, { id: 1 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    expect(wrapper.text()).toContain('+2 Filiais incluídas');
  });

  it('Should render correct border color for status sent documentation', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'in_progress',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    const card = wrapper.findComponent(BCard);
    expect(card.attributes('bordervariant')).toBe('light');
  });

  it('Should render correct border color for status disapproved', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'disapproved',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    const card = wrapper.findComponent(BCard);
    expect(card.attributes('bordervariant')).toBe('danger');
  });

  it('Should render correct border color for status pending documentation', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'pending_documentation',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    const card = wrapper.findComponent(BCard);
    expect(card.attributes('bordervariant')).toBe('light');
  });

  it('Should render correct bg color for status pending documentation', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'pending_documentation',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    const card = wrapper.findComponent(BCard);
    expect(card.classes('pending-documentation-bg-color')).toBeTruthy();
  });

  it('Should show action button when the status is pending_documentation', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'pending_documentation',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    const button = wrapper.findComponent(BButton);
    expect(button.exists()).toBe(true);
  });

  it('Should not show action button when the status is not pending_documentation', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'disapproved',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    const button = wrapper.findComponent(BButton);
    expect(button.exists()).toBe(false);
  });

  it('Should move to economic group data page', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'pending_documentation',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const mockRouter = {
      push: jest.fn(),
    };

    const goToEconomicGroupPage = jest.spyOn(EconomicGroupInformationCard.methods, 'goToEconomicGroupPage');

    const wrapper = await shallowMount(EconomicGroupInformationCard, {
      global: {
        mocks: {
          $router: mockRouter,
        },
      },
      propsData: {
        economicGroup,
      },
    });

    await wrapper.findComponent(BButton).trigger('click');

    expect(goToEconomicGroupPage).toHaveBeenCalled();
    // TODO entender por que o router nao esta sendo inicializado no teste
    // expect(mockRouter.push).toHaveBeenCalledTimes(1)
  });

  it('Should render correct badge style and text when the status is pending_documentation', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'pending_documentation',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    const icon = wrapper.findComponent(FeatherIcon);
    expect(icon.attributes('icon')).toBe('AlertCircleIcon');
    expect(icon.classes()).toContain('text-warning');

    const badge = wrapper.findComponent(BBadge);
    expect(badge.attributes('variant')).toBe('light-warning');
    expect(badge.classes()).toContain('border-warning');
    expect(badge.text()).toBe('Aguardando seu preenchimento');
  });

  it('Should render correct badge style and text when the status is in_progress', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'in_progress',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    const icon = wrapper.findComponent(FeatherIcon);
    expect(icon.attributes('icon')).toBe('ClockIcon');
    expect(icon.classes()).toContain('text-warning');

    const badge = wrapper.findComponent(BBadge);
    expect(badge.attributes('variant')).toBe('light-warning');
    expect(badge.classes()).toContain('border-warning');
    expect(badge.text()).toBe('Informações em análise');
  });

  it('Should render correct badge style and text when the status is disapproved', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'disapproved',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const wrapper = shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });

    const icon = wrapper.findComponent(FeatherIcon);
    expect(icon.attributes('icon')).toBe('AlertOctagonIcon');
    expect(icon.classes()).toContain('text-danger');

    const badge = wrapper.findComponent(BBadge);
    expect(badge.attributes('variant')).toBe('light-danger');
    expect(badge.classes()).toContain('border-danger');
    expect(badge.text()).toBe('Pendência no cadastro');
  });

  xit('Should render EconomicGroupInformationCards correctly for sent status', async () => {
    const economicGroup = {
      id: 1,
      name: 'Teste 1',
      cnpj: '12.345.678/0001-34',
      status: 'pending_documentation',
      filiais: [{ id: 1 }, { id: 2 }, { id: 3 }],
    };

    const wrapper = await shallowMount(EconomicGroupInformationCard, {
      propsData: {
        economicGroup,
      },
    });
  });
});
