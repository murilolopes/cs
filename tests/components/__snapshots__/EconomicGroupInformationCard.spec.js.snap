// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EconomicGroupInformationCard.vue Should render correctly 1`] = `
<b-card-stub
  bodytag="div"
  bordervariant="light"
  class="pending-documentation-bg-color"
  footertag="div"
  headertag="div"
  subtitletag="h6"
  subtitletextvariant="muted"
  tag="div"
  titletag="h4"
>
  <div
    class="d-flex justify-content-between align-items-center mb-50"
  >
    <h3
      class="mb-0"
    >
      Teste 1
    </h3>
     
    <div>
      <feather-icon-stub
        badgeclasses="badge-primary"
        class="text-warning"
        icon="AlertCircleIcon"
        size="26"
      />
       
      <b-badge-stub
        class="border-warning ml-1"
        tag="span"
        target="_self"
        variant="light-warning"
      >
        Aguardando seu preenchimento
      </b-badge-stub>
    </div>
  </div>
   
  <div
    class="d-flex justify-content-between align-items-center"
  >
    <div>
      <h4>
        12.345.678/0001-34
      </h4>
       
      <span>
        +1 Filial incluída
      </span>
    </div>
     
    <b-button-stub
      tag="button"
      target="_self"
      type="button"
      variant="primary"
    >
      Preencher informações agora
    </b-button-stub>
  </div>
</b-card-stub>
`;
